First version implementation of preload_data and the rest of CLMS data store

b-yogesh opened the pull request:
This PR introduces the initial mechanism of preloading data, including cache management, downloading, and file processing. 

The following classes (components) are responsible for the mechanism:

**CLMS**

- Serves as the main interface to interact with the CLMS API. This class coordinates with the _PreloadData_ class to preload the data into a local filestore.

 **CacheManager**

- Manages the local cache of preloaded data.
- Maintains a dictionary (cache) that maps data_ids to their respective file paths.
- Handles file store from the xcube data store  in a local directory and refreshes the cache when necessary.

**DownloadTaskManager**

- Handles the download process, including managing download requests and checking their statuses.
- Retrieves task statuses based on dataset and file IDs or task IDs, determining whether the download is pending, completed, or cancelled.
- Initiates data downloads in chunks and manages zip file extraction, looking specifically for geo data. Definition of geo data is defined in the function docstring in the notes.


**CLMSAPITokenHandler**

- Handles the creation and refreshing of the CLMS API token given the credentials which can be obtained following the steps ![here](https://eea.github.io/clms-api-docs/authentication.html)

**FileProcessor**

- Handles the postprocessing of downloaded data, extracting, stacking and storing geo files from downloaded zip files.

**PreloadData**

- The main class responsible for orchestrating the preloading of datasets.
- It coordinates with _CacheManager_, _DownloadTaskManager_, _CLMSAPITokenHandler_ and _FileProcessor_ classes to handle the complete process of caching, data downloading, making sure token is valid and post-processing of downloaded data.
- Utilizes threading for handling multiple data preloading tasks concurrently.
- Uses `notebook.tqdm` for displaying progress bars

Currently, only logs are shown in the jupyter cells that let the user know of the request status which is updated every 60 seconds to avoid sending several requests to the CLMS API. This is configurable, so it can be changed if we need a lower or higher waiting time between request status. `tqdm` is used here that shows the progress bar for the actual download, unzipping, and postprocessing.

A tutorial notebook has also been added to the `examples` folder to show how to use this data store.

There are some missing tests for a few classes, but they will be added in the next PRs

Closes https://github.com/xcube-dev/xcube-clms/issues/3, https://github.com/xcube-dev/xcube-clms/issues/4, https://github.com/xcube-dev/xcube-clms/issues/5

b-yogesh committed:
Implement initial methods in clms.py

b-yogesh committed:
Refactor code

b-yogesh committed:
Refactor code again

b-yogesh committed:
Implement get_data_ids generator

b-yogesh committed:
Tiny typo fix

b-yogesh committed:
Implement has_data

b-yogesh committed:
Initial version of describe_data (not working entirely)

b-yogesh committed:
Implement describe_data

b-yogesh committed:
Implement get_open_data_params_schema

b-yogesh committed:
Implemented access token class

b-yogesh committed:
Implemented access token class

b-yogesh committed:
[In progress] - open_data implementation

b-yogesh committed:
[In progress] - open_data implementation - implemented _prepare_download_request

b-yogesh committed:
[In progress] - open_data implementation - more impl.

b-yogesh committed:
[In progress] - open_data implementation - more impl.

b-yogesh committed:
Update make_api_request

b-yogesh committed:
temporary bbox and crs handling

b-yogesh committed:
fix condition

b-yogesh committed:
fix get_data_store_params_schema

b-yogesh committed:
Add constants

b-yogesh committed:
Refactoring

b-yogesh committed:
Add schema for preload_data

b-yogesh committed:
Remove error message truncation

b-yogesh committed:
Update get_metadata method

b-yogesh committed:
Add initial unsupported datasets check

b-yogesh committed:
Raise exceptions for unsupported data and add preload params schema

b-yogesh committed:
Update schema methods

b-yogesh committed:
Add TODOs

b-yogesh committed:
Implement first TODO: change data_id def

b-yogesh committed:
Implement first TODO: add include_attr bool impl

b-yogesh committed:
Fix has_data and describe_data

b-yogesh committed:
[WIP] queue download refactor

b-yogesh committed:
[WIP] refactor existing code into preload and clms classes

b-yogesh committed:
[WIP] implement queue downloads.

b-yogesh committed:
[WIP] Further impl. preload

b-yogesh committed:
[WIP] Further impl. preload

b-yogesh committed:
[WIP] Working download impl.

b-yogesh committed:
[WIP] Add initial cancel handler

b-yogesh committed:
[WIP] Add initial cancel handler

b-yogesh committed:
[WIP] Refactoring

b-yogesh committed:
[WIP] Improved download data

b-yogesh committed:
[WIP] Improved download data + merging

b-yogesh committed:
Finish implementation

b-yogesh committed:
Added tests for utils.py

b-yogesh committed:
Added tests for api_token.py

b-yogesh committed:
Refactor clms.py

b-yogesh committed:
Add docstrings clms.py

konstntokas commented on xcube_clms/preload.py:
```diff
@@ -0,0 +1,530 @@
+)
+
+
+class PreloadData:
```
remember Normans issue: https://github.com/xcube-dev/xcube/issues/1093

In the same sense we need to come up with a general set of methods, which defines the PreloadData class. 
So can you sort this out a little bit. This will be helpful when we want to abstract this to an abstract class for xcube.core.store. I suggest, in the most general sense this class should have the following methods: 

- `download_data`: which does the data download. Issuing a request in CLMS is specific to xcube-clms and should go into local methods starting with `_*`. The download data method can take a while, and thus shall support some progress viewing, which you already have. 
- `process_data`: this opens the data, and does the stacking in your case. 
- `monitor_cache`: this will give the user an overview, which datasets are prepared.

Are there more methods which are essential for this class?

b-yogesh replied:
Yeah, I wanted to do this and have done this now.
I think for now, these methods are sufficient. Later, we will need to handle the TaskHandlers that allow the user to view the status of each request and the user can interact with it like cancelling the request.

konstntokas reviewed (changes requested):
Some general notes. 

- Can you filter through, if really everything is needed, or if things can be made simpler. 
- Create a jupyter notebook as an example which shows how the store works. 
- Explain the preload method in the ReadMe

I will continue with the review on Monday.

b-yogesh committed:
Add tests for clms.py

b-yogesh committed:
add preload const

b-yogesh committed:
Refactor

b-yogesh committed:
Add Docstrings and Type Annotations

b-yogesh commented on environment.yml:
```diff
@@ -6,7 +6,10 @@ dependencies:
   - python>=3.10
   - xarray
   - xcube >= 1.7.0
+  - cryptography
```
This is required by the JWT library to create the grant which is required to get the access token for the CLMS API.

b-yogesh commented on xcube_clms/api_token.py:
```diff
@@ -0,0 +1,61 @@
+import time
```
Yeah, I missed in a few. Have added to all the files now. Let me know if I missed any.

b-yogesh commented on xcube_clms/clms.py:
```diff
@@ -18,16 +18,263 @@
+            _, file_id = data_id.split(DATA_ID_SEPARATOR)
+        except ValueError as e:
+            raise ValueError(
+                f"Expected a data_id in the format {{ product_id}}"
```
I think it is already meaningful stating that the user did not provide the right data_id in the format that it is expected. The user can see in the list_data_ids how the data_ids look like to get an idea. I will add that in the example jupyter notebook.

konstntokas replied:
Oh I think I understand your intention better now. I think `format { product_id}` is confusing, except it is a term of your store which is introduced in the ReadMe or documentation.
I just see that you try to split the data_id. So would it not be better to say something that the format of the data ID is wrong?

b-yogesh commented on xcube_clms/clms.py:
```diff
@@ -18,16 +18,263 @@
+
+        return preload_params
+
+    def get_preload_data_params_schema_for_data(self, data_id: str):
```
This method has been removed for now as it serves no purpose. This can be added later if required properly.

b-yogesh commented on xcube_clms/clms.py:
```diff
@@ -18,16 +18,263 @@
+                os.path.join(data_id, data_id_file[0]), **open_params
+            )
+
+    def get_data_ids(
```
They do. But not here. The user will use the store interface and not the CLMS interface. In the store interface, I return an iterator.

konstntokas replied:
In the store interface, a list of all data IDs will be created by calling `get_data_ids` from the `CLMS` class. This can take very long, if a lot of datasets are available within one data store. An iterator only makes sense, when the actual work is performed during iteration. 

I think you just need to exchange a small piece in `CLMS._create_data_ids`. Here you create a list within a for-loop. Instead return a iterator here as well.

b-yogesh commented on xcube_clms/clms.py:
```diff
@@ -18,16 +18,263 @@
+        self._preload_data.initiate_preload(data_id_maps)
+        LOG.info(f"A local Filestore is created at the path: {self.path}")
+
+    def _create_data_ids(
```
See my reply for the same comment

konstntokas replied:
see  my reply to the reply above ;)

b-yogesh commented on xcube_clms/clms.py:
```diff
@@ -18,16 +18,263 @@
+
+        return dict(time_range=time_range, crs=crs[0])
+
+    def get_preload_params(self, data_id: str) -> dict[str : str | None]:
```
I have it turned on, but not sure how sometimes these things are missed.

b-yogesh committed:
Add example notebook

b-yogesh committed:
Minor fixes

b-yogesh committed:
Add missing docs

konstntokas commented:
Something is going wrong in the testing. I created an environment based on `environment.yml` and ran `pytest`. This is the output of the first test failure. All tests show the same error. So I guess it is just one single issue.  

```
(xcube-clms) konstantin@konstantin-nb:~/bc_kon/01_coding/01_github/xcube-clms$ pytest --cov=xcube_clms --cov-report=term-missing
========================================== test session starts ==========================================
platform linux -- Python 3.12.8, pytest-8.3.4, pluggy-1.5.0
rootdir: /home/konstantin/bc_kon/01_coding/01_github/xcube-clms
configfile: pyproject.toml
plugins: recording-0.13.1, cov-6.0.0
collected 38 items                                                                                      

test/test_api_token.py ....                                                                       [ 10%]
test/test_clms.py FF.FFFFFF                                                                       [ 34%]
test/test_utils.py ....................F....Exception ignored in: <function tqdm.__del__ at 0x79b3d9229260>
Traceback (most recent call last):
  File "/home/konstantin/micromamba/envs/xcube-clms/lib/python3.12/site-packages/tqdm/std.py", line 1148, in __del__
    self.close()
  File "/home/konstantin/micromamba/envs/xcube-clms/lib/python3.12/site-packages/tqdm/notebook.py", line 279, in close
    self.disp(bar_style='danger', check_delay=False)
    ^^^^^^^^^
AttributeError: 'tqdm_notebook' object has no attribute 'disp'
                                                      [100%]

=============================================== FAILURES ================================================
_______________________________________ TestCLMS.test_access_item _______________________________________
../../../../micromamba/envs/xcube-clms/lib/python3.12/unittest/mock.py:1393: in patched
    with self.decoration_helper(patched,
../../../../micromamba/envs/xcube-clms/lib/python3.12/contextlib.py:137: in __enter__
    return next(self.gen)
../../../../micromamba/envs/xcube-clms/lib/python3.12/unittest/mock.py:1375: in decoration_helper
    arg = exit_stack.enter_context(patching)
../../../../micromamba/envs/xcube-clms/lib/python3.12/contextlib.py:526: in enter_context
    result = _enter(cm)
../../../../micromamba/envs/xcube-clms/lib/python3.12/unittest/mock.py:1467: in __enter__
    original, local = self.get_original()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <unittest.mock._patch object at 0x79b33771e390>

    def get_original(self):
        target = self.getter()
        name = self.attribute
    
        original = DEFAULT
        local = False
    
        try:
            original = target.__dict__[name]
        except (AttributeError, KeyError):
            original = getattr(target, name, DEFAULT)
        else:
            local = True
    
        if name in _builtins and isinstance(target, ModuleType):
            self.create = True
    
        if not self.create and original is DEFAULT:
>           raise AttributeError(
                "%s does not have the attribute %r" % (target, name)
            )
E           AttributeError: <module 'xcube_clms.preload' from '/home/konstantin/bc_kon/01_coding/01_github/xcube-clms/xcube_clms/preload.py'> does not have the attribute 'CLMSAPIToken'

../../../../micromamba/envs/xcube-clms/lib/python3.12/unittest/mock.py:1437: AttributeError

```

b-yogesh committed:
Fix tests

b-yogesh committed:
Update error message

b-yogesh committed:
Modify list to iterator in get_data_ids

b-yogesh commented:
> Something is going wrong in the testing. I created an environment based on `environment.yml` and ran `pytest`. This is the output of the first test failure. All tests show the same error. So I guess it is just one single issue.
> 
> ```
> (xcube-clms) konstantin@konstantin-nb:~/bc_kon/01_coding/01_github/xcube-clms$ pytest --cov=xcube_clms --cov-report=term-missing
> ========================================== test session starts ==========================================
> platform linux -- Python 3.12.8, pytest-8.3.4, pluggy-1.5.0
> rootdir: /home/konstantin/bc_kon/01_coding/01_github/xcube-clms
> configfile: pyproject.toml
> plugins: recording-0.13.1, cov-6.0.0
> collected 38 items                                                                                      
> 
> test/test_api_token.py ....                                                                       [ 10%]
> test/test_clms.py FF.FFFFFF                                                                       [ 34%]
> test/test_utils.py ....................F....Exception ignored in: <function tqdm.__del__ at 0x79b3d9229260>
> Traceback (most recent call last):
>   File "/home/konstantin/micromamba/envs/xcube-clms/lib/python3.12/site-packages/tqdm/std.py", line 1148, in __del__
>     self.close()
>   File "/home/konstantin/micromamba/envs/xcube-clms/lib/python3.12/site-packages/tqdm/notebook.py", line 279, in close
>     self.disp(bar_style='danger', check_delay=False)
>     ^^^^^^^^^
> AttributeError: 'tqdm_notebook' object has no attribute 'disp'
>                                                       [100%]
> 
> =============================================== FAILURES ================================================
> _______________________________________ TestCLMS.test_access_item _______________________________________
> ../../../../micromamba/envs/xcube-clms/lib/python3.12/unittest/mock.py:1393: in patched
>     with self.decoration_helper(patched,
> ../../../../micromamba/envs/xcube-clms/lib/python3.12/contextlib.py:137: in __enter__
>     return next(self.gen)
> ../../../../micromamba/envs/xcube-clms/lib/python3.12/unittest/mock.py:1375: in decoration_helper
>     arg = exit_stack.enter_context(patching)
> ../../../../micromamba/envs/xcube-clms/lib/python3.12/contextlib.py:526: in enter_context
>     result = _enter(cm)
> ../../../../micromamba/envs/xcube-clms/lib/python3.12/unittest/mock.py:1467: in __enter__
>     original, local = self.get_original()
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <unittest.mock._patch object at 0x79b33771e390>
> 
>     def get_original(self):
>         target = self.getter()
>         name = self.attribute
>     
>         original = DEFAULT
>         local = False
>     
>         try:
>             original = target.__dict__[name]
>         except (AttributeError, KeyError):
>             original = getattr(target, name, DEFAULT)
>         else:
>             local = True
>     
>         if name in _builtins and isinstance(target, ModuleType):
>             self.create = True
>     
>         if not self.create and original is DEFAULT:
> >           raise AttributeError(
>                 "%s does not have the attribute %r" % (target, name)
>             )
> E           AttributeError: <module 'xcube_clms.preload' from '/home/konstantin/bc_kon/01_coding/01_github/xcube-clms/xcube_clms/preload.py'> does not have the attribute 'CLMSAPIToken'
> 
> ../../../../micromamba/envs/xcube-clms/lib/python3.12/unittest/mock.py:1437: AttributeError
> ```

Thanks @konstntokas for the review. I have updated the tests and also refactored the code a bit based on your comments.

b-yogesh committed:
Update CLMSDataStoreTutorial.ipynb

b-yogesh committed:
Update README.md

b-yogesh committed:
Create unittest.yml

b-yogesh committed:
Add test_cache_manager.py

b-yogesh committed:
Add test_token_handler.py

konstntokas commented on .github/unittest.yml:
```diff
@@ -0,0 +1,29 @@
+name: Unittest xcube-clms
```
So far the CI is not executed. I think it is because the unittest.yml needs to be in a folder called workflows. Please double check with other xcube plugin or xcube itself.

konstntokas commented on .github/unittest.yml:
```diff
@@ -0,0 +1,29 @@
+        run: |
+          pytest --cov=xcube_clms --cov-report=xml
+
+      - name: Upload coverage reports to Codecov
```
Let's see if this works out f the box. I think something needs to be set up on codecov webpage, then using the first time.

konstntokas commented on xcube_clms/store.py:
```diff
@@ -98,3 +146,16 @@ def get_search_params_schema(
+
+    @classmethod
+    def get_preload_params_schema(cls):
+        params = dict(
```
`data_ids` is not in `preload_parms`. But you could add a flag if visualization is shown or not. I'd say default to True. In notebook, visualization is useful, but in a script it can be annoying. 
Then you can already validate to the preload-params schema when calling `preload_data`.

konstntokas commented on xcube_clms/token_handler.py:
```diff
@@ -0,0 +1,53 @@
+from xcube_clms.constants import LOG
+
+
+class TokenHandler:
```
Why do you need this, when you already have api_token.py?

b-yogesh replied:
Removed

konstntokas commented:
pytest
========================================== test session starts ==========================================
platform linux -- Python 3.12.8, pytest-8.3.4, pluggy-1.5.0
rootdir: /home/konstantin/bc_kon/01_coding/01_github/xcube-clms
configfile: pyproject.toml
plugins: recording-0.13.1, cov-6.0.0
collected 44 items                                                                                      

test/test_api_token.py ....                                                                       [  9%]
test/test_cache_manager.py ...                                                                    [ 15%]
test/test_clms.py .........                                                                       [ 36%]
test/test_token_handler.py ...                                                                    [ 43%]
test/test_utils.py ....................F....                                                      [100%]

=============================================== FAILURES ================================================
____________________________________ test_cleanup_dir_deletes_files _____________________________________

tmp_path = PosixPath('/tmp/pytest-of-konstantin/pytest-0/test_cleanup_dir_deletes_files0')

    def test_cleanup_dir_deletes_files(tmp_path):
        folder_path = tmp_path / "test_folder"
        folder_path.mkdir()
    
        keep_file = folder_path / "file1.zarr"
        delete_file = folder_path / "file2.tif"
        keep_file.write_text("test")
        delete_file.write_text("test")
    
>       cleanup_dir(folder_path, keep_extension=".tif")

test/test_utils.py:291: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
xcube_clms/utils.py:252: in cleanup_dir
    for filename in tqdm(
../../../../micromamba/envs/xcube-clms/lib/python3.12/site-packages/tqdm/notebook.py:234: in __init__
    self.container = self.status_printer(self.fp, total, self.desc, self.ncols)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

_ = <tqdm.utils.DisableOnWriteError object at 0x72dac7fe8f80>, total = 2
desc = 'Cleaning up directory /tmp/pytest-of-konstantin/pytest-0/test_cleanup_dir_deletes_files0/test_folder'
ncols = None

    @staticmethod
    def status_printer(_, total=None, desc=None, ncols=None):
        """
        Manage the printing of an IPython/Jupyter Notebook progress bar widget.
        """
        # Fallback to text bar if there's no total
        # DEPRECATED: replaced with an 'info' style bar
        # if not total:
        #    return super(tqdm_notebook, tqdm_notebook).status_printer(file)
    
        # fp = file
    
        # Prepare IPython progress bar
        if IProgress is None:  # #187 #451 #558 #872
>           raise ImportError(WARN_NOIPYW)
E           ImportError: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html

../../../../micromamba/envs/xcube-clms/lib/python3.12/site-packages/tqdm/notebook.py:108: ImportError
======================================== short test summary info ========================================
FAILED test/test_utils.py::test_cleanup_dir_deletes_files - ImportError: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readt...
===================================== 1 failed, 43 passed in 4.69s ======================================
Exception ignored in: <function tqdm.__del__ at 0x72db691d85e0>
Traceback (most recent call last):
  File "/home/konstantin/micromamba/envs/xcube-clms/lib/python3.12/site-packages/tqdm/std.py", line 1148, in __del__
    self.close()
  File "/home/konstantin/micromamba/envs/xcube-clms/lib/python3.12/site-packages/tqdm/notebook.py", line 279, in close
    self.disp(bar_style='danger', check_delay=False)
    ^^^^^^^^^
AttributeError: 'tqdm_notebook' object has no attribute 'disp'

One unittest fails

konstntokas commented:
FYI, I guess you can address this in a later PR: 


```
pytest --cov=xcube_clms --cov-report term-missing

---------- coverage: platform linux, python 3.12.8-final-0 -----------
Name                             Stmts   Miss  Cover   Missing
--------------------------------------------------------------
xcube_clms/__init__.py               2      0   100%
xcube_clms/api_token.py             35      0   100%
xcube_clms/cache_manager.py         17      0   100%
xcube_clms/clms.py                  82      6    93%   120, 132, 191, 213-220, 298
xcube_clms/constants.py             80      5    94%   109-115
xcube_clms/download_manager.py     156    135    13%   114-181, 196-214, 233-248, 275-337, 352-439, 462-477
xcube_clms/plugin.py                 5      0   100%
xcube_clms/preload.py               64     34    47%   92-102, 121-165, 181, 190, 196
xcube_clms/processor.py             52     37    29%   68-85, 101-106, 120-159
xcube_clms/store.py                 52     52     0%   21-161
xcube_clms/token_handler.py         12      0   100%
xcube_clms/utils.py                113     10    91%   62-63, 78, 250, 255-262
xcube_clms/version.py                1      0   100%
--------------------------------------------------------------
TOTAL                              671    279    58%
```

b-yogesh committed:
Rename .github/unittest.yml to .github/workflows/unittest.yml

konstntokas commented on environment.yml:
```diff
@@ -6,7 +6,10 @@ dependencies:
   - python>=3.10
   - xarray
```
please also add in `pyproject.toml`

b-yogesh committed:
Add test_processor.py

b-yogesh committed:
Merge remote-tracking branch 'origin/yogesh_preload-data' into yogesh_preload-data

b-yogesh committed:
Update env

b-yogesh committed:
Remove redundant api class

b-yogesh committed:
Update pyproject.toml

b-yogesh committed:
add ipywidgets

b-yogesh committed:
fix test_clms.py

b-yogesh committed:
Add CLMS url as constant

b-yogesh committed:
Fix tests

b-yogesh committed:
Update README.md

b-yogesh committed:
Update README.md

b-yogesh committed:
Update CLMSDataStoreTutorial.ipynb

b-yogesh committed:
Update .gitignore

b-yogesh committed:
Add numpy for tests

b-yogesh committed:
Add missing license text

forman commented on xcube_clms/constants.py:
```diff
@@ -18,5 +18,114 @@
+TASK_STATUS_ENDPOINT = "@datarequest_search"
+CANCEL_ENDPOINT = "@datarequest_delete"
+
+BATCH = "batching"
+NEXT = "next"
```
Such kind of context-free constant definitions do not make much sense.
Declare locally where used or not use at all or add a context prefix.

b-yogesh replied:
cc44bdb17

forman commented on xcube_clms/constants.py:
```diff
@@ -18,5 +18,114 @@
+
+# Dict keys
+
+DATA_ID_KEY = "data_id"
```
Same for the following. Declare isolated where used.

b-yogesh replied:
cc44bdb17

forman commented on xcube_clms/download_manager.py:
```diff
@@ -0,0 +1,477 @@
+from tqdm.notebook import tqdm
+
+from xcube_clms.api_token_handler import CLMSAPITokenHandler
+from xcube_clms.constants import (
```
If the following constants are only used here, declare them here. Use underscore to prevent import from other packages.

b-yogesh replied:
cc44bdb17

forman commented on xcube_clms/store.py:
```diff
@@ -18,75 +18,119 @@
+from .utils import assert_valid_data_type
 
 
 class CLMSDataStore(DataStore, ABC):
```
```suggestion
class ClmsDataStore(DataStore, ABC):
```

forman commented on xcube_clms/store.py:
```diff
@@ -98,3 +142,12 @@ def get_search_params_schema(
     ) -> JsonObjectSchema:
         pass
+
+    def preload_data(self, *data_ids: str, **preload_params):
```
Missing preload handle. See related xcube issue.

b-yogesh replied:
Will be handled in #10

forman commented on xcube_clms/utils.py:
```diff
@@ -0,0 +1,262 @@
+                error_details = response.json()
+            raise HTTPError(f"HTTP error {response.status_code}: {error_details}")
+
+    except JSONDecodeError as e:
+        raise JSONDecodeError(f"Invalid JSON: {e}", response.text, 0)
+    except HTTPError as eh:
+        raise HTTPError(f"HTTP error occurred: {eh}")
+    except Timeout as et:
+        raise Timeout(f"Timeout error occurred: {et}")
+    except RequestException as e:
+        raise RequestException(f"Request error occurred: {e}")
+    except Exception as e:
+        raise Exception(f"Unknown error occurred: {e}")
```
This doesn't make sense. It is just a message mapping - for what reason? You also delete the original traceback.

b-yogesh replied:
f178dce5c0
b30329c9c

forman commented on xcube_clms/utils.py:
```diff
@@ -0,0 +1,262 @@
+# The MIT License (MIT)
```
It seems that the functions in this file help implementing very specific tasks. If such functions are not used twice please define them in the module where they are used. `get_authorization_header` and `spinner` do not have much in common and belong to different contexts.

b-yogesh replied:
b637c7714

forman commented on xcube_clms/processor.py:
```diff
@@ -0,0 +1,162 @@
+from xcube_clms.utils import find_easting_northing, cleanup_dir
+
+
+class FileProcessor:
```
Extract interface and use instead.

forman reviewed (changes requested):
### Store Implementation 

- The preload mechanism lacks the preload handle concept (but that's ok for time being)
- Do not limit the caching data store to `file`. Allow for any `MutableDataStore` instead. 

### Coding

- Using camel-case names, do not concatenate upper-case abbreviations, use camel-case instead: `CLMSAPITokenHandler` is not user friendly, use `ClmsApiTokenHandler` instead
- Don't use `os`, use `fsspec` instead. Actually, don't use specific local filesystem calls at all, if not urgently required. 

### Testing

- Test code is often hard to read because of the bloating mock setup
- Use mocking more carefully! It can cause architectural code changes to be very expensive in the end!
- Test with respect to correct results or behavious, not implementation.
- Build test class names like so: `${class_or_func}Test`, not `Test${class_or_func}`
- Using plain `assert` doesn't distinguish between expected and actual values - which makes it harder to understand what went wrong if tests fail. Prefer test classes that use assert methods from `unitest.TestCase`.

### Docstrings

- Should start with a short, one-line sentence that follows immediately the  `"""` and ends with a dot `.`. Use active speech in this sentence.
- Void methods are supposed to return `None`. Don't document this.


FYI @konstntokas

b-yogesh committed:
Apply suggestions from code review

Co-authored-by: Norman Fomferra <norman.fomferra@brockmann-consult.de>

b-yogesh committed:
Rename classes and fix tests

b-yogesh commented on environment.yml:
```diff
@@ -6,7 +6,13 @@ dependencies:
+  - tqdm
+  - ipywidgets
   # for testing
+  - numpy
```
Currently, I only use it in the tests, so I have not added it in the project dependency.

b-yogesh commented on pyproject.toml:
```diff
@@ -28,8 +34,12 @@ exclude = ["test*", "doc*"]
 
 [project.optional-dependencies]
 dev = [
+    "numpy",
```
I only use `numpy` in the tests, so I have not added it in the `dependencies` list. By others, do you mean the pytest, black, flake8 ... etc.?

b-yogesh commented on xcube_clms/cache_manager.py:
```diff
@@ -0,0 +1,89 @@
+        self.path = path
+        os.makedirs(self.path, exist_ok=True)
+        LOG.info(f"Local Filestore for preload cache created at {self.path}")
+        self.file_store = new_data_store("file", root=self.path)
```
Created an issue #8. This will be taken care of in the next PRs

b-yogesh committed:
Convert file_store and cache to properties

b-yogesh commented on xcube_clms/cache_manager.py:
```diff
@@ -0,0 +1,89 @@
+        """
+        return self.file_store
+
+    def get_cache(self) -> dict[str, str]:
```
78c8943c1102e

b-yogesh committed:
Remove test_store.py

b-yogesh commented on xcube_clms/cache_manager.py:
```diff
@@ -0,0 +1,89 @@
+        """
+        self.cache = None
+        self.path = path
+        os.makedirs(self.path, exist_ok=True)
```
Will be handled later in issue #9

b-yogesh committed:
Remove None return doc

b-yogesh commented on xcube_clms/clms.py:
```diff
@@ -18,16 +18,307 @@
+
+        Returns:
+            None.
+
```
54e2f7fb2

b-yogesh committed:
Improve docstrings

b-yogesh committed:
Update README.md

b-yogesh committed:
Move functions away from utils to respective files

b-yogesh committed:
Move constants to their respective classes

b-yogesh committed:
Improve make_api_request

b-yogesh committed:
Improve make_api_request #2

b-yogesh committed:
Improve tests

b-yogesh committed:
Datastore to MutableDataStore

b-yogesh commented on test/test_clms.py:
```diff
@@ -18,12 +18,545 @@
+from xcube_clms.constants import DATA_ID_SEPARATOR
+
+
+class TestCLMS(unittest.TestCase):
```
bb57381d0

b-yogesh commented on test/test_cache_manager.py:
```diff
@@ -0,0 +1,97 @@
+
+            mock_makedirs.assert_called_once_with(temp_dir, exist_ok=True)
+            mock_new_data_store.assert_called_once_with("file", root=temp_dir)
+            mock_listdir.assert_called_once_with(temp_dir)
```
bb57381d0

b-yogesh commented on test/test_cache_manager.py:
```diff
@@ -0,0 +1,97 @@
+from xcube_clms.cache_manager import CacheManager
+
+
+class TestCacheManager(unittest.TestCase):
```
bb57381d0

konstntokas reviewed (approved):
Approved (since the preload api will change) with some comments which you may address or note down for a later PR.

b-yogesh committed:
Remove init comments

b-yogesh commented on xcube_clms/constants.py:
```diff
@@ -20,86 +20,33 @@
 PENDING = "PENDING"
 COMPLETE = "COMPLETE"
-UNDEFINED = "UNDEFINED"
 CANCELLED = "CANCELLED"
```
Based on what I understand on Norman's comment was to let the constants be here if they are used in multiple files, which is the case for these.

b-yogesh commented on xcube_clms/utils.py:
```diff
@@ -116,26 +139,34 @@ def make_api_request(
-        raise Exception(f"Unknown error occurred: {e}")
+        LOG.error(f"Unknown error occurred: {e}")
+        raise
+
```
That can be done but if I dont do this, the actual error message is lost leaving the user confused about the reason for the exception.
I have added a comment as well explaining the reason I do this.

b-yogesh commented on xcube_clms/utils.py:
```diff
@@ -19,31 +19,25 @@
-    KEEP_EXTENSION,
 )
 
+_PORTAL_TYPE = {"portal_type": "DataSet"}
+_METADATA_FIELDS = "metadata_fields"
+_FULL_SCHEMA = "fullobjects"
```
Yes. I would like all the hardcoded strings to be defined as constants for easier manipulation later (if required).

b-yogesh commented on xcube_clms/clms.py:
```diff
@@ -50,51 +39,57 @@
     ) -> None:
-        """
-        Initialize the CLMS class with API URL, credentials, and optional path.
+        """Initializes the class.
```
27ab2ef

b-yogesh commented:
Closing this PR as #12 refactors this PR.
